### PR TITLE
feat(scoop-info): Show app installed/download size

### LINKED ISSUE #2381: Feature request - Show possible download size in `status` command
When doing a > scoop list command, it would be very helpful to show what apps have updates available, next to it's version number.  Yes, I know that doing scoop update shows new available, but when having many apps installed it's hard to keep track of all that, and the main update command only shows the most recent anyway, not ones it listed in the past.

Also, a upgrade command, like what apt-get does would help:
> scoop upgrade

much like the way apt-get upgrade works, it would be nice for scoop to get the list of upgradable apps, then have it show which ones will upgrade and the amount of data it may take.

Oh, speaking of, it may be helpful to actually put the file sizes in the scoop buckets and have the install command confirm the download first, showing the file size.  I have a data cap, so every byte counts.

### LINKED ISSUE #4840: [Feature] Display app's installation size in `scoop info`
#### Is your feature request related to a problem? Please describe.

I would like to be able to quickly find out how much space a currently installed application is taking on my computer while I am looking at the app's info in `scoop info`. Especially useful if I've just run out of space and want to check if I really need a certain app.

#### Describe the solution you'd like

When running `scoop info` another line would display the app's current size including all versions and any cached downloads. Something like the following:
```
❯ scoop info losslesscut

Name        : losslesscut
Description : Lossless trimming tool for video and audio files
Version     : 3.44.0
Bucket      : extras
...
App size    : 378 MB (471 MB including cache)
```
This would be useful for working out whether an app is worth keeping on smaller hard drives and might remind users to clean their cache.

However the drawback might be a slightly slower output for this command on all types of drives but especially HDDs. Probably negligible for most drives as it's just checking the size of two folders (app folder & cached items).

#### Describe alternatives you've considered

The user could always navigate to the app directory in File Explorer or run WizTree or a similar program but that would require multiple extra steps/programs. This could be good for staying in the terminal while installing/uninstalling old programs and running out of space for example.

### PR DESCRIPTION
**Pull request by tech189**
#### Description

- Adds two new pieces of information to the `scoop info` command when using the `-v` verbose switch
- If the app is installed, it will get the list of files inside the app's installation folder, its persist folder and any cached downloads.
- Then it will add these up and display them one by one with a total underneath.
- If the app isn't installed, then it will try to get the URL from the manifest for the current architecture and get the `Content-Length` header from the download server. It will add up the total for each file given in the manifest.
- Then it will display the total download size or 'Unknown' if the server doesn't provide the info (e.g. sourceforge).
- If the app is cached but not installed, it will let the user know if the latest version is in the cache.

For example, for installed apps:
```
Name           : sharex
Description    : Screen capture, file sharing and productivity tool.
<other properties...>
Installed size : App: 16.1 MB
                 Persist: 768.3 KB
                 Cache: 7.2 MB
                 Total: 24.1 MB
Binaries       : ShareX.exe
Shortcuts      : ShareX
```

Another example with a not yet installed app:
```
Name          : vlc
Description   : A free and open source cross-platform multimedia player and framework that plays most multimedia files
                as well as DVDs, Audio CDs, VCDs, and various streaming protocols.
<other properties...>
Download size : 35.2 MB
Binaries      : vlc.exe
Shortcuts     : VLC media player
```

An example with an app that is uninstalled but the latest version can be installed using cache:
```
Name          : chef-workstation
Description   : Chef Workstation is Chef's modern developer tool kit that includes Chef Infra, InSpec and Habitat plus
                a host of resources, helpers and testing tools that make automating infrastructure, application and
                security testing easier than ever. Chef Workstation supercedes ChefDK.
<other properties...>
Download size : 620.7 MB (latest version is cached)
Shortcuts     : Chef Workstation
Path Added    : C:\Users\me\scoop\apps\chef-workstation\current\bin
```

#### Motivation and Context

- Closes #4840
- Closes #2381
- Allows users to quickly determine the total size of an app without leaving their terminal session
- Allows users to determine the data usage of downloading an app before installing

#### How Has This Been Tested?

- Tested with many apps: those with and without persisted data or cached data, with those that are uninstalled and not.
- Tested to see if it double counts persisted data when evaluating app size - it doesn't.
- Tested the download size with manifests with architecture blocks and those without.
- Tested download size with manifests with multiple urls and those with just one.

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have updated the documentation accordingly.
  - added to changelog
- [ ] I have updated the tests accordingly.
  - not sure which tests should be run

### CODE REVIEW

**Comment by rashil2000** [comment]
Hey, this looks great, but I think you forgot to use the $verbose switch 😅

**Comment by rashil2000** [comment]
There's a catch here. When showing installed size, the code shows the sizes for previous app versions too. This is a bit wrong IMO, since the 'App size' will be double or triple (if there are 2/3 older versions kept) the actual size. One solution is to show another category 'Previous versions' and show their sizes separately from the 'current' version size.

File: libexec/scoop-info.ps1 [hunk 1]
```diff
@@ -115,6 +115,81 @@ if ($status.installed) {
...
+
+        $totalSize = 0
+        $fileTotals = @()
+        foreach ($fileType in ($appFiles, $persistFiles, $cacheFiles)) {
+            if ($fileType.Length -ne 0) {
+                $fileSum = (Get-ChildItem $fileType -Recurse | Measure-Object -Property Length -Sum).Sum
+                $fileTotals += $fileSum
+                $totalSize += $fileSum
+            }
+            else {
```
Comment (rashil2000) [inline libexec/scoop-info.ps1:135]:
```suggestion
            } else {
```

File: libexec/scoop-info.ps1 [hunk 2]
```diff
@@ -115,6 +115,81 @@ if ($status.installed) {
...
+                $fileTotals += 0
+            }
+        }
+
+        $item.'Installed size' = "App: $(filesize $fileTotals[0])`nPersist: $(filesize $fileTotals[1])`nCache: $(filesize $fileTotals[2])`nTotal: $(filesize $totalSize)"
+    }
+}
+else {
```
Comment (rashil2000) [inline libexec/scoop-info.ps1:143]:
```suggestion
} else {
```

File: libexec/scoop-info.ps1 [hunk 3]
```diff
@@ -115,6 +115,81 @@ if ($status.installed) {
...
+        if(!(supports_architecture $manifest $architecture)) {
+        # No available download for current architecture
+            continue
+        }
+
+        if ($null -eq $manifest.url) {
+            # use url for current architecture
+            $urls = url $manifest $architecture
+        }
+        else {
```
Comment (rashil2000) [inline libexec/scoop-info.ps1:157]:
```suggestion
        } else {
```

File: libexec/scoop-info.ps1 [hunk 4]
```diff
@@ -115,6 +115,81 @@ if ($status.installed) {
...
+        $totalPackage = 0
+        foreach($url in $urls) {
+            try {
+                [int]$urlLength = (Invoke-WebRequest $url -Method Head).Headers.'Content-Length'[0]
+                $totalPackage += $urlLength
+
+                if (Test-Path (fullpath (cache_path $app $manifest.version $url))) {
+                    $cached = " (latest version is cached)"
+                }
+                else {
```
Comment (rashil2000) [inline libexec/scoop-info.ps1:171]:
```suggestion
                } else {
```

File: libexec/scoop-info.ps1 [hunk 5]
```diff
@@ -115,6 +115,81 @@ if ($status.installed) {
...
+            catch {
+                $totalPackage = 0
+                $packageError = "the server at $(([System.Uri]$url).Host) is down"
+                break
+            }
+        }
+        if ($totalPackage -ne 0) {
+            $item.'Download size' = "$(filesize $totalPackage)$cached"
+        }
+        else {
```
Comment (rashil2000) [inline libexec/scoop-info.ps1:189]:
```suggestion
        } else {
```

File: libexec/scoop-info.ps1 [hunk 6]
```diff
@@ -115,6 +115,81 @@ if ($status.installed) {
...
+                $totalPackage += $urlLength
+
+                if (Test-Path (fullpath (cache_path $app $manifest.version $url))) {
+                    $cached = " (latest version is cached)"
+                }
+                else {
+                    $cached = $null
+                }
+            }
+            catch [System.Management.Automation.RuntimeException] {
```
Comment (rashil2000) [inline libexec/scoop-info.ps1:175]:
```suggestion
            } catch [System.Management.Automation.RuntimeException] {
```

File: libexec/scoop-info.ps1 [hunk 7]
```diff
@@ -115,6 +115,81 @@ if ($status.installed) {
...
+                $totalPackage = 0
+                $packageError = "the server at $(([System.Uri]$url).Host) did not send a Content-Length header"
+                break
+            }
+            catch {
```
Comment (rashil2000) [inline libexec/scoop-info.ps1:180]:
```suggestion
            } catch {
```

**Comment by tech189** [comment]
Now it looks like this:
```
Installed size : Current version: 70.8 MB
                 Old versions: 139.5 MB
                 Persisted data: 8.1 MB
                 Cached downloads: 50.7 MB
                 Total: 269.1 MB
```

I had to make the category names a bit longer to make them line up roughly. Would it be a good idea to add further spaces so that they line up perfectly like so:

```
Installed size : Current version:  70.8 MB
                 Old versions:     139.5 MB
                 Persisted data:   8.1 MB
                 Cached downloads: 50.7 MB
                 Total:            269.1 MB
```

Or maybe a '-----' divider before the total? Or keep it as it is?

**Comment by rashil2000** [comment]
> Would it be a good idea to add further spaces so that they line up perfectly

Yes!

**Comment by tech189** [comment]
Okay, sorted. Now each total lines up in a column :)

File: libexec/scoop-info.ps1 [hunk 8]
```diff
@@ -143,7 +143,7 @@ if ($status.installed) {
         # Old versions = app total - current version size
         $fileTotals += $fileTotals[0] - $currentTotal
 
-        $item.'Installed size' = "Current version: $(filesize $currentTotal)`nOld versions: $(filesize $fileTotals[3])`nPersisted data: $(filesize $fileTotals[1])`nCached downloads: $(filesize $fileTotals[2])`nTotal: $(filesize $totalSize)"
+        $item.'Installed size' = "Current version:   $(filesize $currentTotal)`nOld versions:      $(filesize $fileTotals[3])`nPersisted data:    $(filesize $fileTotals[1])`nCached downloads:  $(filesize $fileTotals[2])`nTotal:             $(filesize $totalSize)"
```
Comment (rashil2000) [inline libexec/scoop-info.ps1:146]: Split this string into into multiple lines

Comment (tech189) [reply libexec/scoop-info.ps1]: Good idea, done!

File: libexec/scoop-info.ps1 [hunk 9]
```diff
@@ -115,6 +115,87 @@ if ($status.installed) {
...
+
+        # Old versions = app total - current version size
+        $fileTotals += $fileTotals[0] - $currentTotal
+
+        $item.'Installed size' = (
+            "Current version:   $(filesize $currentTotal)`n" +
+            "Old versions:      $(filesize $fileTotals[3])`n" +
+            "Persisted data:    $(filesize $fileTotals[1])`n" +
+            "Cached downloads:  $(filesize $fileTotals[2])`n" +
+            "Total:             $(filesize $totalSize)"
```
Comment (rashil2000) [inline libexec/scoop-info.ps1:151]:
```suggestion
            "Current version:  $(filesize $currentTotal)`n" +
            "Old versions:     $(filesize $fileTotals[3])`n" +
            "Persisted data:   $(filesize $fileTotals[1])`n" +
            "Cached downloads: $(filesize $fileTotals[2])`n" +
            "Total:            $(filesize $totalSize)"
```

Extra space

**Comment by rashil2000** [comment]
Idea: Should we omit a particular size if its value is `0 B`?

**Comment by tech189** [comment]
Output is now as short as possible 🚀

For app with files in each category:
```
Installed size : Current version:  41.2 MB
                 Old versions:     35.4 MB
                 Persisted data:   882 B
                 Cached downloads: 20.1 MB
                 Total:            96.6 MB
```

App with no old versions:
```
Installed size : Current version:  311.4 MB
                 Persisted data:   21.1 MB
                 Cached downloads: 361.5 MB
                 Total:            693.9 MB
```

App with no persisted data:
```
Installed size : Current version:  13.9 MB
                 Cached downloads: 17.3 MB
                 Total:            31.2 MB
```

App with no cache either:
```
Installed size : 276.6 MB
```

**Review by rashil2000 — APPROVED** [review]

**Comment by rashil2000** [comment]
@niheaven can this feature be released in 2.2.0? It's been sitting for a while.

**Comment by niheaven** [comment]
Will review tomorrow and find something to tweak at a glance.

v0.2.2 fixes `Get-Manifest()`'s bugs and should be released ASAP.

File: libexec/scoop-info.ps1 [hunk 10]
```diff
@@ -105,6 +105,105 @@ if ($status.installed) {
...
+        if ($null -eq $manifest.url) {
+            # use url for current architecture
+            $urls = url $manifest $architecture
+        } else {
+            # otherwise use non-architecture url
+            $urls = $manifest.url
+        }
+
+        $totalPackage = 0
+        foreach($url in $urls) {
```
Comment (niheaven) [inline libexec/scoop-info.ps1:181]:
```suggestion
        foreach ($url in @(url $manifest (default_architecture))) {
```

**Comment by tech189** [comment]
Thanks for the review and tweaks! I think it's ready now, I've tested it all too.

I only found one issue but I think it's not related to my changes - I'll open up an issue after this is merged into develop. (if you install an app with app-name@x.x.x syntax, it does not show up in `scoop-info` as 'installed' - and so instead of showing app size, it will show the download size even though it is already installed)

**Review by niheaven — DISMISSED** [review]

**Review by niheaven — APPROVED** [review]

**Review by niheaven — APPROVED** [review]

**Pull request merged**